### ISSUE TITLE
If a fisher can't fish if the road is blocked with roads. The message "No fish" is wrong in this case.

### ISSUE THREAD

**Issue by bunnybot** [widelands/widelands#6243]
[<img align='left' src='https://codeberg.org/avatars/94062fb4c3e503b6434a88938aca6486630fcdfe4d13f557dc2ef771ba3943d8' alt='frankystone' height='48' />](https://codeberg.org/frankystone)*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/4584)*
*Created on Sat Nov 25 11:42:45 CET 2023 by [**frankystone**](https://codeberg.org/frankystone)*

---

There were some questions in the past regarding fishers send a message "No fish" although a fishbreeder says the water is full of fish. See https://www.widelands.org/forum/topic/5886/
Same problem was reported with water gatherers who's working spots are full of roads or trees. See https://www.widelands.org/forum/topic/5621/

Most of the time the spots in the workarea is blocked with trees or roads.

Since those questions arise from time to time there should a descriptive solution.

**Describe the solution you'd like**
This should be handled like for farmers which reports "No Fields": If no spot is found to gather fish or water the message should be something like "No place found" and an according description, e.g. "This worker can't find a place to do his job. Please check if the workarea is blocked by roads or trees."

**Additional context**
Old bugreports which are closed as wontfix:
https://github.com/widelands/widelands/issues/1314
https://github.com/widelands/widelands/issues/1187

**Implementation steps**
- [ ] ...
- [ ] ...
- [ ] ...

**Comment by bunnybot** [comment]
<img align='left' src='https://codeberg.org/avatars/593dc43e2444c25db9481960e5b693ab' alt='tothxa' height='48' />*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/4584#issuecomment-1352579)*
*On Sat Nov 25 14:56:18 CET 2023, **Tóth András (tothxa)** wrote:*

---

Only tangentially related, but should be fixed at the same time: The Frisian charcoal burner failed message says it has no pits, but the failure is more often due to missing logs.

I'm not sure whether we should make it possible to define multiple failure messages for different failure modes (I believe it's not as trivial as it sounds). It may be just as annoying if a player fixes one just to face another, so in this regard, it may be better to list all possibilities, even if it complicates the message.

**Comment by bunnybot** [comment]
<img align='left' src='https://codeberg.org/avatars/94062fb4c3e503b6434a88938aca6486630fcdfe4d13f557dc2ef771ba3943d8' alt='frankystone' height='48' />*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/4584#issuecomment-1352732)*
*On Sat Nov 25 16:56:07 CET 2023, ** (frankystone)** wrote:*

---

> I'm not sure whether we should make it possible to define multiple failure messages for different failure modes (I believe it's not as trivial as it sounds). It may be just as annoying if a player fixes one just to face another, so in this regard, it may be better to list all possibilities, even if it complicates the message.

I agree. There are probably more failures possible and trying to create one specific message for each failure may blow up the code.

Other opinions?

**Comment by hessenfarmer** [comment]
I believe the problem is that we do only have one type of failure reported by the production program. so it is not only about messages.

**Comment by bunnybot** [comment]
<img align='left' src='https://codeberg.org/avatars/94062fb4c3e503b6434a88938aca6486630fcdfe4d13f557dc2ef771ba3943d8' alt='frankystone' height='48' />*[Mirrored from Codeberg](https://codeberg.org/wl/widelands/issues/4584#issuecomment-1352895)*
*On Sat Nov 25 18:56:40 CET 2023, ** (frankystone)** wrote:*

---

Currently there are two, `kDefault` and `kFull`: https://codeberg.org/wl/widelands/src/commit/38cd8bf0d8a6be3c8228a02927d74eec4ae8d631/src/logic/map_objects/tribes/productionsite.h#L39

Where `kFull` is used if the production site is full of wares and `kDefault` for all other use cases, as far i interpret the code correctly: https://codeberg.org/wl/widelands/src/commit/38cd8bf0d8a6be3c8228a02927d74eec4ae8d631/src/logic/map_objects/tribes/productionsite.cc#L1184

Maybe it would be useful for the AI to have more of such failure status?

**Comment by cake-pie** [comment]
Absent multiple failure modes with different messages, it is possible to alleviate the issue by "overloading" the existing message to cover the various possibilities, as an interim measure. In the case of fishers, add text to remind that fishing is not possible along shoreline that is blocked.

```diff
-"The fisher working out of this fisher’s house can’t find any fish in his work area."
+"The fisher working out of this fisher’s house can’t find any fish in his work area. Note that roads and trees along the shoreline block fishing."
```
Possible disadvantage is the longer message may bit unwieldy for translation; maybe define the note as a common string to concat/append to the existing message to avoid duplication.

As precedent, the atlanteans fisher already includes a reminder about using a breeder to replenish fish stocks.

https://github.com/widelands/widelands/blob/2d0d5d877fbd2d0298824220e0423710009771cc/data/tribes/buildings/productionsites/atlanteans/fishers_house/init.lua#L58
